### PR TITLE
Adding span batch sampler processor (#3435)

### PR DESCRIPTION
**Pull request by markvoll**
* Initial commit for new processor
* Changelog/codeowners/issue_template updates
* config implementation with tests
* factory implementation with tests
* Base level processor implementation without filtration
* filtration interface and methods to be added in next commit

**Description:** <Describe what has changed.>

This is the first commit for a new processor described in the tracking issue below

**Link to tracking Issue:** <Issue number if applicable>
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/15630

**Testing:** <Describe what testing was performed and which tests were added.>
* Initial config unit tests
* Initial factory unit tests

**Documentation:** <Describe the documentation added.>
* README.md file

### CODE REVIEW

**Comment by markvoll** [comment]
FYI @jpkrohling

File: processor/spanbatchsamplerprocessor/go.mod [hunk 1]
```diff
@@ -0,0 +1,41 @@
+module github.com/open-telemetry/opentelemetry-collector-contrib/processor/spanbatchsamplerprocessor
+
+go 1.19
+
+require (
+	github.com/stretchr/testify v1.8.1
+	go.opencensus.io v0.24.0
+	go.opentelemetry.io/collector v0.68.0
```
Comment (jpkrohling) [inline processor/spanbatchsamplerprocessor/go.mod:8]: We are now at 0.70.0, sorry for taking so long to review this.

File: processor/spanbatchsamplerprocessor/testdata/config.yaml [hunk 2]
```diff
@@ -0,0 +1 @@
+tokens: 1000
```
Comment (jpkrohling) [inline processor/spanbatchsamplerprocessor/testdata/config.yaml:1]: The config on the testdata should be a complete config for the component, potentially with different component IDs ("spanbatchsampler/1", "spanbatchsampler/default", ...)

File: processor/spanbatchsamplerprocessor/processor.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+// Shutdown is invoked during service shutdown.
+func (sbf *spanBatchSamplerProcessor) Shutdown(context.Context) error {
+	close(sbf.shutdownC)
+
+	// Wait until all goroutines are done.
+	sbf.goroutines.Wait()
+	return nil
+}
+
+func (sbf *spanBatchSamplerProcessor) addItem(item interface{}) {
```
Comment (jpkrohling) [inline processor/spanbatchsamplerprocessor/processor.go:136]: Does this need to be an interface{}?

File: processor/spanbatchsamplerprocessor/processor.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+		// spanFilter:     filter, // TODO add filter interface
+	}, nil
+}
+
+func newBatchTraces(nextConsumer consumer.Traces) *batchTraces {
+	return &batchTraces{nextConsumer: nextConsumer, traceData: ptrace.NewTraces(), sizer: &ptrace.ProtoMarshaler{}}
+}
+
+func (sbf *spanBatchSamplerProcessor) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{MutatesData: false}
```
Comment (jpkrohling) [inline processor/spanbatchsamplerprocessor/processor.go:117]: This actually mutates data (reorgaized the batches), it should be `true`.

**Review by jpkrohling — COMMENTED** [review]
I wasn't up to speed with the latest proposal, so, let's sort it out before moving forward with the coding.

You mentioned earlier that it would be a sampler of traces, which makes sense. Now, it's a sampler of spans, which is a bit confusing to me: are you sampling only a individual spans from a given trace? Does it mean that traces might be incomplete at the storage layer?

There was also a proposal for a similar component by @yuvalo1212 and we discussed it today during the SIG Collector: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/17874

We agreed today that the proposal above would be folded into the tail-sampling processor. If both components are similar, I would prefer to move forward in having a policy for the tail-sampling processor instead of a new component.

**Comment by clemenskol** [comment]
> I wasn't up to speed with the latest proposal, so, let's sort it out before moving forward with the coding.
> 
> You mentioned earlier that it would be a sampler of traces, which makes sense. Now, it's a sampler of spans, which is a bit confusing to me: are you sampling only a individual spans from a given trace? Does it mean that traces might be incomplete at the storage layer?
> 
> There was also a proposal for a similar component by @yuvalo1212 and we discussed it today during the SIG Collector: #17874
> 
> We agreed today that the proposal above would be folded into the tail-sampling processor. If both components are similar, I would prefer to move forward in having a policy for the tail-sampling processor instead of a new component.

hi @jpkrohling 

thanks for chiming in on this - I was also involved in some of this proposal behind the scenes and had a few questions/comments on your reply

First off, I think the proposal in #17874 is in deed very interesting and applicable to us, but there are a few points in our proposal for which I'd like to understand how it would work in the other proposal:

first, our intention was to use an interface that allows extending the hashing of a single span in a very simple manner, because the definition of how 2 spans are the same could really depend on the use-case. For example: I intended to have a "HTTP-aware span hasher" that would look at the HTTP headers, query parameters, POST body, etc. and use their values to decide if 2 HTTP requests are the same (modulo minor changes, such as timestamps).
This way, we could consider a span identical if they had the same headers and parameters set (with "somewhat similar" header/parameter values), but if 2 requests differ in headers or parameters, they would be considered different.

The linked proposal looks only at trace similarity if all spans match, and span equality is defined on an application-agnostic level. This of course makes sense in some cases, but for us it would not work, as it's not expressive enough.

Do you think there is a plan to make the proposal more powerful in that direction?

second, this proposal has the concept of a "token" that decides how much to upload. This allows *fair* distribution that allows duplicates. The alternative proposal only looks at unique uploads in a configured time (essentially until `deduplication_timeout` removes a hash from the "previously seen traces" map).
In some of our use-cases, this is a problem, as we do want a steady stream of events, even if they are duplicates - as long as we don't get overwhelmed by the uploads.

Last, the focus on spans, rather than traces, was very intentional although it - as you correctly pointed out - means that we end up with partial traces. The problem that the other proposal somewhat side-steps is that of load-balancing. It is very difficult to expect that the data collection is configured in a way that all spans for a trace end up on the same collector (where the in-memory map can be used to filter duplicates).
By operating on the span level, this problem does not exist, because a collector can make a drop-decision based on the information it has, regardless if a trace consists of spans that might be seen by different collector instances.
In our scenarios, having individual spans removed from traces is acceptable (or even desired).

Thanks for your help reviewing this proposal and looking forward to hearing your thoughts!

**Comment by jpkrohling** [comment]
@clemenskol, I would suggest adding this again as a topic to the next SIG Collector meeting.

> The problem that the other proposal somewhat side-steps is that of load-balancing. It is very difficult to expect that the data collection is configured in a way that all spans for a trace end up on the same collector (where the in-memory map can be used to filter duplicates).

We have the load balancing exporter for this. Have you considered including it in the architecture for your final solution? Dropping spans from traces is an entirely new discussion and potentially causes user confusion, as the absence of data is also indicative of a faulty behavior (eg, did this operation not happen, or was it removed?)

**Comment by clemenskol** [comment]
> @clemenskol, I would suggest adding this again as a topic to the next SIG Collector meeting.
> 
> > The problem that the other proposal somewhat side-steps is that of load-balancing. It is very difficult to expect that the data collection is configured in a way that all spans for a trace end up on the same collector (where the in-memory map can be used to filter duplicates).
> 
> We have the load balancing exporter for this. Have you considered including it in the architecture for your final solution?

please excuse the ignorance on my behalf, but it's not clear to me how this helps. Isn't this exporter just useful for data already in opentelemetry collector (for sending upstream)? 

the problem I'm referring to is what happens when you have multiple load-balanced collectors that the proposal you linked kind of side-steps: the caller sending data to otel-collector must make sure data is all routed to the same instance. If that's not the case, the approach breaks.

or are you proposing that we have a "first layer" of otel-collectors that use the router to then send all traces to a "second layer" of otel-collectors and uses the load-balancing to group traces together? We have not considered this as it seems too complicated, but maybe I'm misunderstanding your proposal

> Dropping spans from traces is an entirely new discussion and potentially causes user confusion, as the absence of data is also indicative of a faulty behavior (eg, did this operation not happen, or was it removed?)

You have a point, but I think that really depends on the user expectation. Data could be marked as removed or the trace could be tagged accordingly. 
in our scenario, we really want the span to be dropped to avoid duplication

**Comment by jpkrohling** [comment]
> or are you proposing that we have a "first layer" of otel-collectors that use the router to then send all traces to a "second layer" of otel-collectors and uses the load-balancing to group traces together?

This.

**Comment by clemenskol** [comment]
> > or are you proposing that we have a "first layer" of otel-collectors that use the router to then send all traces to a "second layer" of otel-collectors and uses the load-balancing to group traces together?
> 
> This.

I see. Seems

1) somewhat resource-wasteful to run 2x the instances of the collector to deal with this

2) error prone - having to configure 2 different types of collectors although the user really just want to collect and filter.

But thanks for clarifying, I understand what you propose (although I'm curious if it's really a great solution for users in practice)

**Comment by jpkrohling** [comment]
To clarify: when scaling up a collector, you need a way to send the spans with the same trace_id to the same collector instance. So, you need the load balancing on the first level. On the second level, you apply the sampling, typically employing the tail-sampling processor, which accepts or rejects the entire trace.

**Pull request closed without merging**